src/gadget.rs

```diff
@@ -0,0 +1,175 @@
+//! Self-contained circuit implementations of various primitives.
+
+use std::fmt;
+
+use crate::{
+    arithmetic::FieldExt,
+    plonk::{Advice, Any, Column, ConstraintSystem, Error, Fixed},
+};
+
+pub mod sha256;
```

Comment:
This module should not depend on `sha256`.

Reply:
It doesn't? Nothing in this module depends on this line; it just exposes the `sha256` submodule to crate users.

Reply:
What's special about the `sha256` module that we should publically expose it here, rather than having its users import it directly? IOW I think it should not be a submodule of `gadget`, because user-implemented concrete gadget modules won't be.

Reply:
Am I misunderstanding you? A user can't "import it directly" in Rust. If the module files are located in `halo2::gadget::sha256` (which they currently are), then that module path needs to be exposed in the public API. Alternatively, the files could be moved somewhere else, or re-exported as e.g. `halo::sha256` (which starts polluting the top-level namespace of the `halo2` crate). But there *must* be a `pub mod sha256` somewhere, or else the gadget is unreachable by end users.